rfc/0001-policy-server-custom-resource.md

```diff
@@ -0,0 +1,415 @@
+be served by the `default` one.
+
+
+## Changes to the current Kubewarden controller
+
+Currently, the Kubewarden controller watches the `ClusterAdmissionPolicy` resources.
+The controller will be changed to monitor both `ClusterAdmissionPolicy` and
+`PolicyServer` resources.
+
+However, the controller will implement a reconciliation loop only for the `PolicyServer`
+resource.
+Whenever something happens against a `ClusterAdmissionPolicy`, the controller will
+find the `PolicyServer` resource that owns it, and trigger a reconciliation event
+against that `PolicyServer` instance.
+
+The creation/update of a PolicyServer object leads to the following actions
+inside of the controller:
+
+  * List all the ClusterAdmissionPolicy resources that belong to this PolicyServer
+    instance. Create/Update a ConfigMap that holds the `policies.yml` configuration.
+    We could name this ConfigMap `policy-server-<name of the policy server>`.
+  * Create a Deployment that runs Policy Server: use the information stored into
+    the Custom Resource, plus ensure the `policy-server-<name of the policy server>`
+    ConfigMap is mounted into the deployment
+  * Ensure the Deployment is exposed inside of the cluster via a dedicated
+    Service. We can name the service `policy-server-<name of the policy server>`
+  * For each policy managed by the Policy Server, create/update the
+    ValidatingWebhookConfiguration/MutatingWebhookConfiguration resources
+    associated with them.
+
+Some of the resources mentioned above are Namespace-based (like Deployment, ConfigMap, Service);
+in the beginning all these resources will be created in the same Namespace where the controller
+operates. We might want to make that tunable in the future, but right now it would just be
+an overkill.
+
+When a `ClusterAdmissionPolicy` resource is created/updated:
+
+  * Find the `PolicyServer` that owns the policy
+  * Trigger the reconciliation loop of the Policy Server. This will lead to an update
+    of the ConfigMap that holds the `policies.yaml` file, and a rollout of the Deployment
+
+
+Special handling must be done for deletion events. Both `PolicyServer` and `ClusterAdmissionPolicy`
+resources must have a finalizer set.
+
+Deletion of a `ClusterAdmissionPolicy`:
+
+  * Thanks to the finalizer, the `ClusterAdmissionPolicy` resource is not yet deleted
+    from etcd. We can hence find the `PolicyServer` that owns the about-to-be-deleted
+    policy. We will use this information later on.
+  * Delete the `ValidatingWebhookConfiguration` or the `MutatingWebhookConfiguration`
+    resource associated with the policy.
+  * Delete the `ClusterAdmissionPolicy` from etcd
+  * Perform the reconciliation loop of the `PolicyServer` that owned the policy
+
+Deletion of a `PolicyServer`:
+
+  * Thanks to the finalizer, the `PolicyServer` resource is not yet deleted
+    from etcd.
+  * Go through all the policies that are owned by the `PolicyServer` instance,
```

Comment:
Just learned about [owner references](https://kubernetes.io/blog/2021/05/14/using-finalizers-to-control-deletion/#owner-references). 
Would it make sense for the controller to add owner references to policies with the policy-server as parent? (and maybe do the same for WebhookConfigurations and policies?).

That way things would get scheduled for deletion automatically, and can be a signifier that the controller could find the policy-server that is in control of those policies, as listed in the policies.

Owner references seem to be a list; one can have several, which would make it work for several policy-servers consuming the same policies.

Reply:
We can use the cascade deletion, yes. I would leave it open in the wording of this document and decide in the implementation phase.

Reply:
This is a good point. I was pondering if, by setting ownership relations, we could even avoid to use the finalizers.
I think this cannot be done, because we have the following resources:

|Resource | Scope |
|------------------|-----|
| PolicyServer CR | Cluster |
| Policy Server Service | Namespace |
| Policy Server Deployment | Namespace |
| Policy Server ConfigMap | Namespace |
| Policy Server Secret | Namespace |
| ValidatingWebhookConfiguration/MutatingWebhookConfiguration | Cluster |

I don't know if the Kubernetes GC allows a Cluster-scoped resource to own namespaced ones and, most important of all, if it handles the cleanup properly...

I know there were limitations in the past, but I couldn't find anything right now

Reply:
I've updated the doc to mention the usage of owner references.